File: src/llm/verifier.ts
import { LLMProvider } from '../llm/types';
import { ChatMessage } from '../types/chat';
import { ModelConfig } from '../types/config';

export interface VerificationRequest {
    prompt: string;
    response: string;
    context: string[];
}

export interface VerificationResult {
    approved: boolean;
    confidence: number;
    notes?: string;
}

/**
 * Run a verification pass using a small, strict LLM.
 * This verifier is expected to be deterministic.
 */
export async function verifyWithLLM(
    provider: LLMProvider,
    model: ModelConfig,
    request: VerificationRequest
): Promise<VerificationResult> {
    const messages = buildVerifierMessages(request);

    const result = await provider.chat({
        model,
        messages,
        stream: false,
    });

    return parseVerifierOutput(result.content);
}

/**
 * Construct a strict verification prompt.
 */
function buildVerifierMessages(request: VerificationRequest): ChatMessage[] {
    return [
        {
            role: 'system',
            content:
                'You are a strict verifier. Your job is to approve or reject answers. ' +
                'You must be conservative and reject if unsure.',
        },
        {
            role: 'user',
            content: `
USER PROMPT:
${request.prompt}

MODEL RESPONSE:
${request.response}

REFERENCE CONTEXT:
${request.context.map((c, i) => `[${i + 1}] ${c}`).join('\n')}

TASK:
1. Is the response correct?
2. Is it fully supported by the reference context?
3. Does it avoid speculation or fabrication?

Respond ONLY with valid JSON in the following format:
{
  "approved": boolean,
  "confidence": number,
  "notes": string
}
            `.trim(),
        },
    ];
}

/**
 * Parse verifier output safely.
 */
function parseVerifierOutput(text: string): VerificationResult {
    try {
        const start = text.indexOf('{');
        const end = text.lastIndexOf('}');

        if (start === -1 || end === -1) {
            throw new Error('No JSON block found');
        }

        const parsed = JSON.parse(text.slice(start, end + 1));

        return {
            approved: Boolean(parsed.approved),
            confidence:
                typeof parsed.confidence === 'number'
                    ? clamp(parsed.confidence, 0, 1)
                    : 0,
            notes:
                typeof parsed.notes === 'string'
                    ? parsed.notes
                    : undefined,
        };
    } catch {
        return {
            approved: false,
            confidence: 0,
            notes: 'Verifier output could not be parsed',
        };
    }
}

/**
 * Clamp confidence values.
 */
function clamp(
    value: number,
    min: number,
    max: number
): number {
    return Math.max(min, Math.min(max, value));
}

File: rust/rlm/src/llm/verifier.rs
use crate::llm::types::{LLMChatRequest, LLMProvider};
use crate::types::chat::ChatMessage;
use crate::types::config::ModelConfig;

pub struct VerificationRequest {
    pub prompt: String,
    pub response: String,
    pub context: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct VerificationResult {
    pub approved: bool,
    pub confidence: f64,
    pub notes: Option<String>,
}

pub async fn verify_with_llm<P: LLMProvider + ?Sized>(
    provider: &P,
    model: ModelConfig,
    request: VerificationRequest,
) -> anyhow::Result<VerificationResult> {
    let messages = build_verifier_messages(&request);

    let result = provider
        .chat(LLMChatRequest {
            model,
            messages,
            stream: false,
        })
        .await?;

    Ok(parse_verifier_output(&result.content))
}

fn build_verifier_messages(request: &VerificationRequest) -> Vec<ChatMessage> {
    vec![
        ChatMessage {
            role: crate::types::chat::ChatRole::System,
            content: "You are a strict verifier. Your job is to approve or reject answers. You must be conservative and reject if unsure.".to_string(),
        },
        ChatMessage {
            role: crate::types::chat::ChatRole::User,
            content: format!(
                "USER PROMPT:\n{}\n\nMODEL RESPONSE:\n{}\n\nREFERENCE CONTEXT:\n{}\n\nTASK:\n1. Is the response correct?\n2. Is it fully supported by the reference context?\n3. Does it avoid speculation or fabrication?\n\nRespond ONLY with valid JSON in the following format:\n{{\n  \"approved\": boolean,\n  \"confidence\": number,\n  \"notes\": string\n}}",
                request.prompt,
                request.response,
                request
                    .context
                    .iter()
                    .enumerate()
                    .map(|(i, c)| format!("[{}] {}", i + 1, c))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
        },
    ]
}

fn parse_verifier_output(text: &str) -> VerificationResult {
    let (start, end) = match (text.find('{'), text.rfind('}')) {
        (Some(s), Some(e)) if s <= e => (s, e),
        _ => {
            return VerificationResult {
                approved: false,
                confidence: 0.0,
                notes: Some("Verifier output could not be parsed".to_string()),
            }
        }
    };

    let slice = &text[start..=end];
    match serde_json::from_str::<serde_json::Value>(slice) {
        Ok(parsed) => {
            let approved = parsed
                .get("approved")
                .and_then(|v| v.as_bool())
                .unwrap_or(false);
            let confidence = parsed
                .get("confidence")
                .and_then(|v| v.as_f64())
                .map(|c| clamp(c, 0.0, 1.0))
                .unwrap_or(0.0);
            let notes = parsed
                .get("notes")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());

            VerificationResult {
                approved,
                confidence,
                notes,
            }
        }
        Err(_) => VerificationResult {
            approved: false,
            confidence: 0.0,
            notes: Some("Verifier output could not be parsed".to_string()),
        },
    }
}

fn clamp(value: f64, min: f64, max: f64) -> f64 {
    if value < min {
        min
    } else if value > max {
        max
    } else {
        value
    }
}
